File: OneCodeTeam/Story-Creator-Sample-Application-for-Windows-14992/[VB][C++]-Story Creator Sample Application for Windows Phone 7 (VBWPAzureVideoStory)/C++ and VB/C++,VB/VBWPAzureVideoStory/StoryCreatorWebRole/StoryService.vb
'****************************** Module Header ******************************\
' Module Name:	StoryService.vb
' Project: StoryCreatorWebRole
' Copyright (c) Microsoft Corporation.
' 
' A WCF REST service built with Web API.
' 
' This source is subject to the Microsoft Public License.
' See http://www.microsoft.com/opensource/licenses.mspx#Ms-PL.
' All other rights reserved.
' 
' THIS CODE AND INFORMATION IS PROVIDED "AS IS" WITHOUT WARRANTY OF ANY KIND, 
' EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE IMPLIED 
' WARRANTIES OF MERCHANTABILITY AND/OR FITNESS FOR A PARTICULAR PURPOSE.
'***************************************************************************/



Imports System.ServiceModel
Imports System.Net.Http
Imports System.Net
Imports System.Runtime.Serialization.Json
Imports System.IO
Imports Microsoft.WindowsAzure.StorageClient
Imports System.ServiceModel.Web
Imports System.Xml
Imports StoryDataModel

<ServiceContract()>
Public Class StoryService
    ''' <summary>
    ''' Create a new story resource.
    ''' </summary>
    <WebInvoke(Method:="POST", UriTemplate:="")> _
    Public Function Post(request As HttpRequestMessage) As HttpResponseMessage
        '  Make sure the storage account is ready.
        If Global_asax.StorageAccount Is Nothing Then
            If Not Global_asax.InitializeStorage() Then
                Return Me.CreateStringResponse(HttpStatusCode.BadRequest, "The service is unavailable at the moment. Please try again later.")
            End If
        End If

        Try
            ' Request body.
            Dim docSource As XDocument = XDocument.Parse(request.Content.ReadAsString())

            ' Response body.
            Dim docResult As New XDocument(New XElement("Story"))


            Dim photos = docSource.Root.Elements("Photo")
            Dim photoElementsCount As Integer = photos.Count()
            If photos.Count() < 2 Then
                Return Me.CreateStringResponse(HttpStatusCode.BadRequest, "A story requires at least 2 photos.")
            End If

            Dim photoCount As Integer = 0
            Try
                photoCount = Integer.Parse(docSource.Root.Attribute("PhotoCount").Value)
            Catch
                Return Me.CreateStringResponse(HttpStatusCode.BadRequest, "The request body is not well formatted. The required attribute PhotoCount of element Story is either missing or incorrect.")
            End Try

            If photoElementsCount <> photoCount Then
                Return Me.CreateStringResponse(HttpStatusCode.BadRequest, "The request body is not well formatted. The value of PhotoCount doesn't equal to count of the Photo elements.")
            End If

            Dim blobClient As New CloudBlobClient(Global_asax.StorageAccount.BlobEndpoint, Global_asax.StorageAccount.Credentials)
            Dim container As CloudBlobContainer = blobClient.GetContainerReference("videostories")

            ' The unique ID that represents the story.
            Dim id As String = Guid.NewGuid().ToString()
            Dim configBlob As CloudBlob = container.GetBlobReference(id & ".xml")
            docResult.Root.Add(New XAttribute("ID", id))

            For Each photo As XElement In photos
                Dim name As String = photo.Attribute("Name").Value

                ' Construct SAS. The start time is set to 1 minute earlier,
                ' to make sure the client is able to upload the blob.
                Dim blob As CloudBlob = container.GetBlobReference(id & "/" & name)
                Dim sas As String = blob.GetSharedAccessSignature(New SharedAccessPolicy() With { _
                  .Permissions = SharedAccessPermissions.Write, _
                  .SharedAccessStartTime = DateTime.Now.AddMinutes(-1.0), _
                  .SharedAccessExpiryTime = DateTime.Now.AddHours(0.5) _
                })

                ' Create an empty blob, so clients can upload to the correct blob.
                blob.UploadText("")

                ' Modify the original configuration. Add URI without SAS.
                photo.Add(New XAttribute("Uri", blob.Uri.AbsoluteUri))

                ' Add the Photo element to the response, including SAS.
                Dim fullUri As String = blob.Uri.AbsoluteUri & sas
                docResult.Root.Add(New XElement("Photo", New XAttribute("Name", name), New XAttribute("Uri", fullUri)))
            Next

            ' Store the config in blob storage.
            configBlob.UploadText(docSource.ToString())

            Trace.Write("Story configuraion created: " & configBlob.Uri.ToString(), "Information")

            ' Return the success response.
            Return Me.CreateXmlResponse(HttpStatusCode.Created, docResult.ToString())
        Catch generatedExceptionName As XmlException
            Return Me.CreateStringResponse(HttpStatusCode.BadRequest, "The request body is not a well formatted xml document.")
        Catch ex As StorageClientException
            Trace.Write("Error dealing with blob: " + ex.Message, "Error")
            Return Me.CreateStringResponse(HttpStatusCode.InternalServerError, "The service is unavailable at the moment. Please try again later.")
        End Try
    End Function

    ''' <summary>
    ''' Update a story resource.
    ''' Currently the only update is to commit the story (indicates we can start to encode the video).
    ''' </summary>
    <WebInvoke(Method:="PUT", UriTemplate:="{id}?commit={commit}")> _
    Public Function Put(request As HttpRequestMessage, id As String, commit As System.Nullable(Of Boolean)) As HttpResponseMessage
        ' Make sure the storage account is ready.
        If Global_asax.StorageAccount Is Nothing Then
            If Not Global_asax.InitializeStorage() Then
                Return Me.CreateStringResponse(HttpStatusCode.BadRequest, "The service is unavailable at the moment. Please try again later.")
            End If
        End If
        If String.IsNullOrEmpty(id) Then
            Return Me.CreateStringResponse(HttpStatusCode.BadRequest, "Required parameter id is missing.")
        End If
        If commit Is Nothing OrElse Not commit.Value Then
            Return Me.CreateStringResponse(HttpStatusCode.BadRequest, "Currently only ""commit=true""  option is supported.")
        End If

        Try
            Dim blobClient As New CloudBlobClient(Global_asax.StorageAccount.BlobEndpoint, Global_asax.StorageAccount.Credentials)
            Dim container As CloudBlobContainer = blobClient.GetContainerReference("videostories")
            Dim configBlob As CloudBlob = container.GetBlobReference(id & ".xml")

            ' We do not actually need those attributes. We simply check if the blob exists or not.
            ' If the blob doesn't exist, a StorageClientException will be thrown, so we jump to the catch block.
            configBlob.FetchAttributes()

            ' Add the job to queue.
            Dim queueClient As New CloudQueueClient(Global_asax.StorageAccount.QueueEndpoint, Global_asax.StorageAccount.Credentials)
            Dim queue As CloudQueue = queueClient.GetQueueReference("videostories")
            queue.AddMessage(New CloudQueueMessage(id))

            ' Return an empty successful message.

            Return Me.CreateStringResponse(HttpStatusCode.NoContent, "")
        Catch ex As StorageClientException
            If ex.StatusCode = HttpStatusCode.NotFound Then
                Return Me.CreateStringResponse(HttpStatusCode.NotFound, "The requested story does not exist.")
            End If

            ' General error, trace and return a genric message.
            Trace.Write("Error dealing with blob: " + ex.Message, "Error")
            Return Me.CreateStringResponse(HttpStatusCode.InternalServerError, "The service is unavailable at the moment. Please try again later.")
        End Try
    End Function

    <WebGet(UriTemplate:="")> _
    Public Function [Get](request As HttpRequestMessage) As HttpResponseMessage
        ' Make sure the storage account is ready.
        If Global_asax.StorageAccount Is Nothing Then
            If Not Global_asax.InitializeStorage() Then
                Return Me.CreateStringResponse(HttpStatusCode.BadRequest, "The service is unavailable at the moment. Please try again later.")
            End If
        End If

        Try
            Dim storyDataContext As New StoryDataContext(Global_asax.StorageAccount.TableEndpoint.AbsoluteUri, Global_asax.StorageAccount.Credentials)

            ' Query the table storage.
            Dim query = From s In storyDataContext.Stories

            ' convert the result to a simplified class that doesn't contain partition/row key.
            Dim stories As New List(Of Story)()
            For Each s As StoryDataModel.Story In query
                stories.Add(New Story() With { _
                  .Name = s.Name, _
                  .VideoUri = s.VideoUri _
                })
            Next
            Dim jsonSerializer As New DataContractJsonSerializer(GetType(List(Of Story)))
            Using stream As New MemoryStream()
                jsonSerializer.WriteObject(stream, stories)
                stream.Position = 0
                Using reader As New StreamReader(stream)
                    Dim result As String = reader.ReadToEnd()
                    Return Me.CreateJsonResponse(HttpStatusCode.OK, result)
                End Using
            End Using
        Catch ex As StorageClientException
            If ex.StatusCode = HttpStatusCode.NotFound Then
                Return Me.CreateStringResponse(HttpStatusCode.NotFound, "The requested story does not exist.")
            End If

            ' General error, trace and return a genric message.
            Trace.Write("Error dealing with table service: " + ex.Message, "Error")
            Return Me.CreateStringResponse(HttpStatusCode.InternalServerError, "The service is unavailable at the moment. Please try again later.")
        Catch ex2 As Data.EvaluateException
            Trace.Write("Error dealing with table service: " + ex2.Message, "Error")
            Return Me.CreateStringResponse(HttpStatusCode.InternalServerError, "The service is unavailable at the moment. Please try again later.")
        End Try
    End Function

    Private Function CreateStringResponse(statusCode As HttpStatusCode, body As String) As HttpResponseMessage
        Dim response As New HttpResponseMessage()
        response.StatusCode = statusCode
        response.Content = New StringContent(body)
        Return response
    End Function

    Private Function CreateXmlResponse(statusCode As HttpStatusCode, body As String) As HttpResponseMessage
        Dim response As New HttpResponseMessage()
        response.StatusCode = statusCode
        response.Content = New StringContent(body, Encoding.UTF8, "text/xml")
        Return response
    End Function

    Private Function CreateJsonResponse(statusCode As HttpStatusCode, body As String) As HttpResponseMessage
        Dim response As New HttpResponseMessage()
        response.StatusCode = statusCode
        response.Content = New StringContent(body, Encoding.UTF8, "application/json")
        Return response
    End Function
End Class

File: OneCodeTeam/Story-Creator-Sample-Application-for-Windows-14638/[C#][C++]-Story Creator Sample Application for Windows Phone 7 (CSWP7AzureVideoStory)/C# and C++/C#,C++/CSWP7AzureVideoStory/StoryCreatorWebRole/StoryService.cs
/****************************** Module Header ******************************\
* Module Name:	StoryService.cs
* Project: StoryCreatorWebRole
* Copyright (c) Microsoft Corporation.
* 
* A WCF REST service built with Web API.
* 
* This source is subject to the Microsoft Public License.
* See http://www.microsoft.com/opensource/licenses.mspx#Ms-PL.
* All other rights reserved.
* 
* THIS CODE AND INFORMATION IS PROVIDED "AS IS" WITHOUT WARRANTY OF ANY KIND, 
* EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE IMPLIED 
* WARRANTIES OF MERCHANTABILITY AND/OR FITNESS FOR A PARTICULAR PURPOSE.
\***************************************************************************/

using System;
using System.Collections.Generic;
using System.Data.Services.Client;
using System.Diagnostics;
using System.IO;
using System.Linq;
using System.Net;
using System.Net.Http;
using System.Runtime.Serialization.Json;
using System.ServiceModel;
using System.ServiceModel.Web;
using System.Text;
using System.Xml;
using System.Xml.Linq;
using Microsoft.WindowsAzure.StorageClient;
using StoryDataModel;

namespace StoryCreatorWebRole
{
    [ServiceContract]
    public class StoryService
    {
        /// <summary>
        /// Create a new story resource.
        /// </summary>
        [WebInvoke(Method = "POST", UriTemplate = "")]
        public HttpResponseMessage Post(HttpRequestMessage request)
        {
            // Make sure the storage account is ready.
            if (Global.StorageAccount == null)
            {
                if (!Global.InitializeStorage())
                {
                    return this.CreateStringResponse(HttpStatusCode.BadRequest,
                        "The service is unavailable at the moment. Please try again later.");
                }
            }

            try
            {
                // Request body.
                XDocument docSource = XDocument.Parse(request.Content.ReadAsString());

                // Response body.
                XDocument docResult = new XDocument(new XElement("Story"));
                ;
                var photos = docSource.Root.Elements("Photo");
                int photoElementsCount = photos.Count();
                if (photos.Count() < 2)
                {
                    return this.CreateStringResponse(HttpStatusCode.BadRequest, "A story requires at least 2 photos.");
                }

                int photoCount = 0;
                try
                {
                    photoCount = int.Parse(docSource.Root.Attribute("PhotoCount").Value);
                }
                catch
                {
                    return this.CreateStringResponse(HttpStatusCode.BadRequest,
                        "The request body is not well formatted. The required attribute PhotoCount of element Story is either missing or incorrect.");
                }

                if (photoElementsCount != photoCount)
                {
                    return this.CreateStringResponse(HttpStatusCode.BadRequest,
                        "The request body is not well formatted. The value of PhotoCount doesn't equal to count of the Photo elements.");
                }

                CloudBlobClient blobClient = new CloudBlobClient(Global.StorageAccount.BlobEndpoint, Global.StorageAccount.Credentials);
                CloudBlobContainer container = blobClient.GetContainerReference("videostories");

                // The unique ID that represents the story.
                string id = Guid.NewGuid().ToString();
                CloudBlob configBlob = container.GetBlobReference(id + ".xml");
                docResult.Root.Add(new XAttribute("ID", id));

                foreach (XElement photo in photos)
                {
                    string name = photo.Attribute("Name").Value;

                    // Construct SAS. The start time is set to 1 minute earlier,
                    // to make sure the client is able to upload the blob.
                    CloudBlob blob = container.GetBlobReference(id + "/" + name);
                    string sas = blob.GetSharedAccessSignature(new SharedAccessPolicy()
                    {
                        Permissions = SharedAccessPermissions.Write,
                        SharedAccessStartTime = DateTime.Now.AddMinutes(-1d),
                        SharedAccessExpiryTime = DateTime.Now.AddHours(0.5)
                    });

                    // Create an empty blob, so clients can upload to the correct blob.
                    blob.UploadText("");

                    // Modify the original configuration. Add URI without SAS.
                    photo.Add(new XAttribute("Uri", blob.Uri.AbsoluteUri));

                    // Add the Photo element to the response, including SAS.
                    string fullUri = blob.Uri.AbsoluteUri + sas;
                    docResult.Root.Add(new XElement("Photo", new XAttribute("Name", name), new XAttribute("Uri", fullUri)));
                }

                // Store the config in blob storage.
                configBlob.UploadText(docSource.ToString());

                Trace.Write("Story configuraion created: " + configBlob.Uri, "Information");

                // Return the success response.
                return this.CreateXmlResponse(HttpStatusCode.Created, docResult.ToString());
            }
            catch (XmlException)
            {
                return this.CreateStringResponse(HttpStatusCode.BadRequest,
                    "The request body is not a well formatted xml document.");
            }
            catch (StorageClientException ex)
            {
                Trace.Write("Error dealing with blob: " + ex.Message, "Error");
                return this.CreateStringResponse(HttpStatusCode.InternalServerError,
                    "The service is unavailable at the moment. Please try again later.");
            }
        }

        /// <summary>
        /// Update a story resource.
        /// Currently the only update is to commit the story (indicates we can start to encode the video).
        /// </summary>
        [WebInvoke(Method = "PUT", UriTemplate = "{id}?commit={commit}")]
        public HttpResponseMessage Put(HttpRequestMessage request, string id, bool? commit)
        {
            // Make sure the storage account is ready.
            if (Global.StorageAccount == null)
            {
                if (!Global.InitializeStorage())
                {
                    return this.CreateStringResponse(HttpStatusCode.BadRequest,
                        "The service is unavailable at the moment. Please try again later.");
                }
            }
            if (string.IsNullOrEmpty(id))
            {
                return this.CreateStringResponse(HttpStatusCode.BadRequest, "Required parameter id is missing.");
            }
            if (commit == null || !commit.Value)
            {
                return this.CreateStringResponse(HttpStatusCode.BadRequest, "Currently only \"commit=true\" option is supported.");
            }

            try
            {
                CloudBlobClient blobClient = new CloudBlobClient(Global.StorageAccount.BlobEndpoint, Global.StorageAccount.Credentials);
                CloudBlobContainer container = blobClient.GetContainerReference("videostories");
                CloudBlob configBlob = container.GetBlobReference(id + ".xml");

                // We do not actually need those attributes. We simply check if the blob exists or not.
                // If the blob doesn't exist, a StorageClientException will be thrown, so we jump to the catch block.
                configBlob.FetchAttributes();

                // Add the job to queue.
                CloudQueueClient queueClient = new CloudQueueClient(Global.StorageAccount.QueueEndpoint, Global.StorageAccount.Credentials);
                CloudQueue queue = queueClient.GetQueueReference("videostories");
                queue.AddMessage(new CloudQueueMessage(id));

                // Return an empty successful message.
                return this.CreateStringResponse(HttpStatusCode.NoContent, "");

            }
            catch (StorageClientException ex)
            {
                if (ex.StatusCode == HttpStatusCode.NotFound)
                {
                    return this.CreateStringResponse(HttpStatusCode.NotFound, "The requested story does not exist.");
                }

                // General error, trace and return a genric message.
                Trace.Write("Error dealing with blob: " + ex.Message, "Error");
                return this.CreateStringResponse(HttpStatusCode.InternalServerError,
                    "The service is unavailable at the moment. Please try again later.");
            }
        }

        [WebGet(UriTemplate = "")]
        public HttpResponseMessage Get(HttpRequestMessage request)
        {
            // Make sure the storage account is ready.
            if (Global.StorageAccount == null)
            {
                if (!Global.InitializeStorage())
                {
                    return this.CreateStringResponse(HttpStatusCode.BadRequest,
                        "The service is unavailable at the moment. Please try again later.");
                }
            }

            try
            {
                StoryDataContext storyDataContext = new StoryDataContext(Global.StorageAccount.TableEndpoint.AbsoluteUri, Global.StorageAccount.Credentials);
                
                // Query the table storage.
                var query = from s in storyDataContext.Stories select s;

                // convert the result to a simplified class that doesn't contain partition/row key.
                List<Story> stories = new List<Story>();
                foreach (StoryDataModel.Story s in query)
                {
                    stories.Add(new Story() { Name = s.Name, VideoUri = s.VideoUri });
                }
                DataContractJsonSerializer jsonSerializer = new DataContractJsonSerializer(typeof(List<Story>));
                using (MemoryStream stream = new MemoryStream())
                {
                    jsonSerializer.WriteObject(stream, stories);
                    stream.Position = 0;
                    using (StreamReader reader = new StreamReader(stream))
                    {
                        string result = reader.ReadToEnd();
                        return this.CreateJsonResponse(HttpStatusCode.OK, result);
                    }
                }
            }
            catch (StorageClientException ex)
            {
                if (ex.StatusCode == HttpStatusCode.NotFound)
                {
                    return this.CreateStringResponse(HttpStatusCode.NotFound, "The requested story does not exist.");
                }

                // General error, trace and return a genric message.
                Trace.Write("Error dealing with table service: " + ex.Message, "Error");
                return this.CreateStringResponse(HttpStatusCode.InternalServerError,
                    "The service is unavailable at the moment. Please try again later.");
            }
            catch (DataServiceQueryException ex2)
            {
                Trace.Write("Error dealing with table service: " + ex2.Message, "Error");
                return this.CreateStringResponse(HttpStatusCode.InternalServerError,
                    "The service is unavailable at the moment. Please try again later.");
            }
        }

        private HttpResponseMessage CreateStringResponse(HttpStatusCode statusCode, string body)
        {
            HttpResponseMessage response = new HttpResponseMessage();
            response.StatusCode = statusCode;
            response.Content = new StringContent(body);
            return response;
        }

        private HttpResponseMessage CreateXmlResponse(HttpStatusCode statusCode, string body)
        {
            HttpResponseMessage response = new HttpResponseMessage();
            response.StatusCode = statusCode;
            response.Content = new StringContent(body, Encoding.UTF8, "text/xml");
            return response;
        }

        private HttpResponseMessage CreateJsonResponse(HttpStatusCode statusCode, string body)
        {
            HttpResponseMessage response = new HttpResponseMessage();
            response.StatusCode = statusCode;
            response.Content = new StringContent(body, Encoding.UTF8, "application/json");
            return response;
        }
    }
}